framework/src/engine/legacy/legacy_chain_handler.ts

```diff
@@ -12,56 +12,215 @@
+				// eslint-disable-next-line @typescript-eslint/no-misused-promises
+				this._syncTimeout = setTimeout(async () => {
+					await this._trySyncBlocks(bracket, lastBlock);
+				}, 120000); // 2 mints = (60 * 2) * 1000
+			} else {
+				throw err;
+			}
+		}
+	}
+
+	/**
+	 * Flow of syncing legacy blocks
+	 *
+	 * Check if we have `sync` property `true` in configuration (already done in engine.ts)
+	 * call getConnectedPeers from network
+	 * Filter peers having their `legacy` buffer array contains `snapshotBlockID`
+	 * If there is no peer, throw error, this error will be used in outside function to retry calling `syncBlocks` after x amount of time
+	 * Get a random peer from list of filtered peers with legacy info
+	 * Make a request to that random peer by calling its `getLegacyBlocksFromId` method with `data` property set to `legacyBlock.header.id`
+	 * Try to decode response data buffer using getBlocksFromIdResponseSchema, apply penalty in case of error & retry syncBlocks
+	 * Validate `blocks: Buffer[]`, apply penalty in case of error & retry syncBlocks
+	 * Decode `blocks: Buffer[]` into LegacyBlock[] & start saving one by one
+	 * If last block height is still higher than bracket.startHeight, save bracket with `lastBlockHeight: lastBlock?.header.height` & repeat syncBlocks
+	 * If last block height equals bracket.startHeight, simply save bracket with `lastBlockHeight: lastBlock?.header.height`
+	 */
+	// eslint-disable-next-line @typescript-eslint/member-ordering
+	public async syncBlocks(bracket: LegacyBlockBracket, legacyBlock: LegacyBlock): Promise<void> {
+		const connectedPeers = (this._network.getConnectedPeers() as unknown) as Peer[];
+		const peersWithLegacyInfo = connectedPeers.filter(
+			peer =>
+				!!(peer.options as { legacy: Buffer[] }).legacy.find(snapshotBlockID =>
+					snapshotBlockID.equals(Buffer.from(bracket.snapshotBlockID, 'hex')),
+				),
+		);
+		if (!peersWithLegacyInfo) {
+			throw new PeerNotFoundWithLegacyInfo('No peer found with legacy info.');
+		}
+
+		const randomPeerIndex = Math.trunc(Math.random() * peersWithLegacyInfo.length - 1);
+		const { peerId } = peersWithLegacyInfo[randomPeerIndex];
+
+		const p2PRequestPacket: P2PRequestPacket = {
+			procedure: 'getLegacyBlocksFromId',
+			data: legacyBlock.header.id,
+		};
+
+		const response = await this._network.requestFromPeer({ ...p2PRequestPacket, peerId });
+
+		// `data` is expected to hold blocks in DESC order
+		const { data } = response;
+		let legacyBlocks: LegacyBlock[];
+
+		const applyPenaltyAndRepeat = async () => {
+			this._network.applyPenaltyOnPeer({ peerId, penalty: 100 });
+			await this.syncBlocks(bracket, legacyBlock);
+		};
+
+		try {
+			// this part is needed to make sure `data` returns ONLY `{ blocks: Buffer[] }` & not any extra field(s)
+			const { blocks } = codec.decode<{ blocks: Buffer[] }>(
+				getBlocksFromIdResponseSchema,
+				data as Buffer,
+			);
+			if (!(blocks.length > 0)) {
```

Comment:
For this ` if (!(blocks.length > 0)) ` piece of logic, would it be better to use the opposite operator (<=) `if (blocks.length <= 0)` instead? Not a show stopper by any means, just a suggestion.